codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoWriter.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import java.util.function.BiFunction;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.utils.CodeWriter;
+
+/**
+ * Specialized code writer for managing Go dependencies.
+ *
+ * <p>Use the {@code $T} formatter to refer to {@link Symbol}s.
+ */
+public final class GoWriter extends CodeWriter {
+
+    private final String fullPackageName;
+
+    public GoWriter(String fullPackageName) {
+        this.fullPackageName = fullPackageName;
+        trimBlankLines();
+        trimTrailingSpaces();
+        setIndentText("\t");
+        putFormatter('T', new GoSymbolFormatter());
+    }
+
+    @Override
+    public String toString() {
+        String contents = super.toString();
+        String[] packageParts = fullPackageName.split("/");
+        String header = String.format(
+                "// Code generated by smithy-go-codegen DO NOT EDIT.%npackage %s%n%n",
+                packageParts[packageParts.length - 1]);
+        // TODO: add imports
+        return header + contents;
```

Comment:
Don't have to do it ATM but wanted to call out, that it would be good to consider how we'll generate [summary/overview documentation API packages](https://blog.golang.org/godoc).

The package doc string should only appear in one file within a package, (e.g. doc.go file) and provides the documentation/summary/overview of a package. We can determine if this should be a canned string, or incorporate documentation provided by the API model, (e.g [v1 sdk's ACM](https://docs.aws.amazon.com/sdk-for-go/api/service/acm/#pkg-overview))

```go
// Code generated by smithy-go-codegen DO NOT EDIT.

// Package weather provides the client, and types for the Weather API. <- first paragraph is summary
//
// ... could be more that only appears in API docs overview, not included in package's summary.
package weather
```

Reply:
Eventually we'll add a `useFileWriter` to `GoDelegator`, which we'll then use to write out such files. We should include the service shape's documentation there where available.